pkg/metallb/frr.go

```diff
@@ -0,0 +1,559 @@
+
+		builder.errorMsg = "frrConfiguration 'remoteAS' cannot be empty when defining receive all mode"
+	}
+
+	// Defining the neighbor with peer IP and ASN information.
+	if len(builder.Definition.Spec.BGP.Routers[0].Neighbors) == 0 {
+		builder.Definition.Spec.BGP.Routers[0].Neighbors = []frrtypes.Neighbor{{
+			Address:   bgpPeerIP,
+			ASN:       remoteAS,
+			ToReceive: frrtypes.Receive{Allowed: frrtypes.AllowedInPrefixes{Mode: frrtypes.AllowAll}},
+		},
+		}
+	} else {
+		builder.Definition.Spec.BGP.Routers[0].Neighbors = append(builder.Definition.Spec.BGP.Routers[0].Neighbors,
+			frrtypes.Neighbor{
+				Address:   bgpPeerIP,
+				ASN:       remoteAS,
+				ToReceive: frrtypes.Receive{Allowed: frrtypes.AllowedInPrefixes{Mode: frrtypes.AllowAll}},
+			})
+	}
+
+	return builder
+}
+
+// WithToReceiveModeFiltered allows only specified prefixes to be received.
+func (builder *FrrConfigurationBuilder) WithToReceiveModeFiltered(bgpPeerIP string, remoteAS uint32,
+	prefixes []string) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Allows the FRR nodes to receive specific routes in namespace %s in mode type: filtered",
+		builder.Definition.Namespace)
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	if bgpPeerIP == "" {
+		glog.V(100).Infof("The BGP peer IP should not be blank")
+
+		builder.errorMsg = fmt.Sprintf("the BGP Peer IP %s is an empty string", bgpPeerIP)
+	}
+
+	if strconv.Itoa(int(remoteAS)) == "" {
+		glog.V(100).Infof("The remoteASN of the frrConfiguration can not be empty")
+
+		builder.errorMsg = "frrConfiguration 'remoteAS' cannot be empty"
+	}
+
+	// Defining the neighbor with peer IP and ASN information.
+	neighbor := frrtypes.Neighbor{
+		Address: bgpPeerIP,
+		ASN:     remoteAS,
+		ToReceive: frrtypes.Receive{
+			Allowed: frrtypes.AllowedInPrefixes{
+				Prefixes: make([]frrtypes.PrefixSelector, len(prefixes)),
+			},
+		},
+	}
+
+	// Iterate the prefix list.
+	for i, prefix := range prefixes {
+		glog.V(100).Infof("AllowMode 'filtered' receives only routes within the prefix %v", prefix)
+		neighbor.ToReceive.Allowed.Prefixes[i] = frrtypes.PrefixSelector{Prefix: prefix}
+	}
+
+	builder.Definition.Spec.BGP.Routers[0].Neighbors = append(builder.Definition.Spec.BGP.Routers[0].Neighbors, neighbor)
+
+	return builder
+}
+
+// WithBGPNeighbor defines a single neighbor IP and ASN number.
+func (builder *FrrConfigurationBuilder) WithBGPNeighbor(bgpPeerIP string, remoteAS uint32) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Define BGP Neighbor %s with peer IP %s and ASN %v",
+		builder.Definition.Name, bgpPeerIP, remoteAS)
+
+	if bgpPeerIP == "" {
+		glog.V(100).Infof("The BGP peer IP should not be blank")
+
+		builder.errorMsg = fmt.Sprintf("the BGP peer IP %s is an empty string", bgpPeerIP)
+	}
+
+	if net.ParseIP(bgpPeerIP) == nil {
```

Comment:
Do we need to add this bgpPeerIp validation to the WithToReceiveModeFiltered()  and  WithToReceiveModeAll()?

Reply:
yes added